Runtime/Services/BacktraceSession.cs

```diff
@@ -0,0 +1,317 @@
+        /// <param name="timeIntervalInMs">Update time interval in MS</param>
+        /// <param name="maximumNumberOfEventsInStore">Determine how many events we can store in event store</param>
+        public BacktraceSession(
+            AttributeProvider attributeProvider,
+            string uploadUrl,
+            long timeIntervalInMs,
+            uint maximumNumberOfEventsInStore)
+        {
+            SubmissionUrl = uploadUrl;
+            _attributeProvider = attributeProvider;
+            _maximumNumberOfEventsInStore = maximumNumberOfEventsInStore;
+            _timeIntervalInMs = timeIntervalInMs;
+            RequestHandler = new BacktraceHttpClient();
+        }
+
+        /// <summary>
+        /// Backtrace session update interval method used by Backtrace client to update session events time
+        /// </summary>
+        /// <param name="time">Current game time</param>
+        public void Tick(long time)
+        {
+            if (_timeIntervalInMs == 0)
+            {
+                return;
+            }
+            lock (_object)
+            {
+                var intervalUpdate = (time - _lastUpdateTime) >= _timeIntervalInMs;
+                var reachedEventLimit = _maximumNumberOfEventsInStore == Count() && _maximumNumberOfEventsInStore != 0;
+                if (intervalUpdate == false && reachedEventLimit == false)
+                {
+                    // nothing more to update
+                    return;
+                }
+                _lastUpdateTime = time;
+            }
+            Send();
+        }
+
+        /// <summary>
+        /// Force BacktraceSession to Send data to Backtrace
+        /// </summary>
+        public void Send()
+        {
+            Send(numberOfRetries: 3);
+        }
+
+        /// <summary>
+        /// Force BacktraceSession to Send data to Backtrace with retry settings
+        /// </summary>
+        public void Send(int numberOfRetries = DefaultNumberOfRetries)
+        {
+            if (numberOfRetries == 0)
+            {
+                Debug.LogWarning("Backtrace Session: Cannot send session data to Backtrace due to submission issue.");
+                CleanEvents();
+                return;
+            }
+            if (Count() == 0)
+            {
+                return;
+            }
+            // create a copy of events 
+            var uniqueEvents = UniqueEvents.ToArray();
+            var sessionEvents = SessionEvents.ToArray();
+            var payload = CreateJsonPayload(uniqueEvents, sessionEvents);
+
+            // cleanup existing copy of events
+            CleanEvents();
+
+            // submit data to Backtrace
+            RequestHandler.Post(SubmissionUrl, payload, (long statusCode, bool httpError, string response) =>
+            {
+                if (statusCode == 200)
+                {
+                    OnRequestCompleted();
+                }
+                else

```

Comment:
On first read, it seems that this doesn't try with backoff.

Reply:
Since we're not in API that allows us to use coroutines (mono behavior) we can't, for example, wait for n seconds without blocking the main thread. We shouldn't also use threads because we previously agreed that this will increase resources that backtrace-unity integration consumes and might impact games with a small pc budget. unfortunately, I can't have a better solution for exponential backoff than trying instantly just after the previous request :(

Reply:
Given the volume in question, we can't afford for these to spin out if the service goes down.  We can wait for @wca to chime in if we want but this will effectively end up DDOSing ourselves one day.  if we have to use a thread, then we probably have to use a thread.  Maybe we can only do it when automatic submission is enabled?

I see that the interface for requests uses a callback; do we have a way to also do callbacks on timers, e..g JS setTimeout or something?

Reply:
Hey. I've moved this code to tick/update method. Tick/Update method will be invoked by Untiy engine every time when unity generates a new frame. So tick method will occur every time when unity will try to render a new frame. With that being said without an additional thread/timer we can use the update method to schedule jobs to be sent by BacktraceSession.

Now I added default timeout which is 10 seconds. Every time when we failed we're waiting at least 10 seconds for the next retry. Also like previously - we're making only 3 requests. 

@ahicks92 what do you think about this solution?